sdk/src/main/scala/io/horizen/account/sc2sc/AbstractCrossChainRedeemMessageProcessor.scala

```diff
@@ -0,0 +1,150 @@
+package io.horizen.account.sc2sc
+
+import com.google.common.primitives.Bytes
+import io.horizen.account.state.events.AddCrossChainRedeemMessage
+import io.horizen.account.state.{BaseAccountStateView, ExecutionRevertedException, Message, NativeSmartContractMsgProcessor}
+import io.horizen.cryptolibprovider.{CryptoLibProvider, Sc2scCircuit}
+import io.horizen.params.NetworkParams
+import io.horizen.sc2sc.{CrossChainMessage, CrossChainMessageHash}
+import io.horizen.utils.BytesUtils
+import sparkz.crypto.hash.Keccak256
+
+import java.nio.charset.StandardCharsets
+
+trait CrossChainRedeemMessageProvider {
+  def doesCrossChainMessageHashFromRedeemMessageExist(hash: CrossChainMessageHash, view: BaseAccountStateView): Boolean
+}
+
+abstract class AbstractCrossChainRedeemMessageProcessor(
+                                                         networkParams: NetworkParams,
+                                                         sc2scCircuit: Sc2scCircuit,
+                                                         scTxMsgProc: ScTxCommitmentTreeRootHashMessageProvider
+                                                       ) extends NativeSmartContractMsgProcessor with CrossChainRedeemMessageProvider {
+  protected def processRedeemMessage(msg: AccountCrossChainRedeemMessage, view: BaseAccountStateView): Array[Byte] = {
+    validateRedeemMsg(msg, view)
+
+    addCrossChainMessageToView(view, msg)
+
+    addCrossChainRedeemMessageLogEvent(view, msg)
+
+    msg.encode()
+  }
+
+  private def addCrossChainRedeemMessageLogEvent(view: BaseAccountStateView, accCcRedeemMessage: AccountCrossChainRedeemMessage): Unit = {
+    val event = AddCrossChainRedeemMessage(
+      accCcRedeemMessage.accountCrossChainMessage,
+      accCcRedeemMessage.certificateDataHash,
+      accCcRedeemMessage.nextCertificateDataHash,
+      accCcRedeemMessage.scCommitmentTreeRoot,
+      accCcRedeemMessage.nextScCommitmentTreeRoot,
+      accCcRedeemMessage.proof
+    )
+    val evmLog = getEthereumConsensusDataLog(event)
+    view.addLog(evmLog)
+  }
+
+  private def addCrossChainMessageToView(view: BaseAccountStateView, accCcRedeemMessage: AccountCrossChainRedeemMessage): Unit = {
+    val messageHash = CryptoLibProvider.sc2scCircuitFunctions.getCrossChainMessageHash(
```

Comment:
```suggestion
    val messageHash = sc2scCircuit.getCrossChainMessageHash(
```
don't use `CryptoLibProvider.sc2scCircuitFunctions`.
And in general the same comment as in UTXO model: let's use OOP style: `accCcRedeemMessage.getAccountCrooschainMessage.getHash`